// iotex-antenna/lib/antenna.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _contract = require("./contract/contract");

var _iotx = require("./iotx");

var _ws = require("./plugin/ws");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Antenna {
  constructor(provider, opts) {
    _defineProperty(this, "iotx", void 0);

    this.iotx = new _iotx.Iotx(provider, {
      signer: opts && opts.signer,
      timeout: opts && opts.timeout,
      apiToken: opts && opts.apiToken
    });
  }

  setProvider(provider) {
    if (typeof provider === "object") {
      if (provider === this.iotx.currentProvider()) {
        return;
      }
    }

    this.iotx.setProvider(provider);
  }

  currentProvider() {
    return this.iotx.currentProvider();
  }

}

exports.default = Antenna;

_defineProperty(Antenna, "modules", {
  Iotx: _iotx.Iotx,
  WsSignerPlugin: _ws.WsSignerPlugin,
  Contract: _contract.Contract
});
