// iotex-antenna/lib/account/accounts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Accounts = void 0;

var _account = _interopRequireDefault(require("eth-lib/lib/account"));

var _account2 = require("./account");

var _wallet = _interopRequireDefault(require("./wallet"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Accounts extends Array {
  constructor() {
    super();

    _defineProperty(this, "wallet", void 0);

    this.wallet = new _wallet.default();
    return new Proxy(this, {
      get: (target, name) => {
        // @ts-ignore
        if (target.wallet[name]) {
          // @ts-ignore
          return target.wallet[name];
        } // @ts-ignore


        return target[name];
      }
    });
  }

  create(entropy) {
    const acct = _account.default.create(entropy);

    const privateKey = acct.privateKey.substr(2);

    const realAccount = _account2.Account.fromPrivateKey(privateKey);

    this.wallet.add(realAccount);
    return realAccount;
  }

  privateKeyToAccount(privateKey) {
    const account = _account2.Account.fromPrivateKey(privateKey);

    this.wallet.add(account);
    return account;
  }

  addressToAccount(address) {
    const account = _account2.Account.fromAddress(address);

    this.wallet.add(account);
    return account;
  }

  addAccount(account) {
    this.wallet.add(account);
    return account;
  }

  getAccount(address) {
    // @ts-ignore
    return this.wallet[address];
  }

  removeAccount(address) {
    return this.wallet.remove(address);
  }

  async sign(data, privateKey) {
    return _account2.Account.fromPrivateKey(privateKey).sign(data);
  }

}

exports.Accounts = Accounts;
