// iotex-antenna/lib/account/account.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RemoteAccount = exports.Account = void 0;

var _crypto = require("../crypto/crypto");

var _hash = require("../crypto/hash");

var _utils = require("./utils");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Account {
  constructor() {
    _defineProperty(this, "address", void 0);

    _defineProperty(this, "privateKey", void 0);

    _defineProperty(this, "publicKey", void 0);
  }

  static fromPrivateKey(privateKey) {
    if (privateKey === "") {
      throw new Error("private key can not empty.");
    }

    const obj = (0, _crypto.privateKeyToAccount)(privateKey);
    const act = new Account();
    act.address = obj.address;
    act.privateKey = obj.privateKey;
    act.publicKey = obj.publicKey;
    return act;
  }

  static fromAddress(address) {
    if (address === "" || address.length !== 41) {
      throw new Error("address illegal.");
    }

    const act = new Account();
    act.address = address;
    return act;
  }

  async sign(data) {
    if (!this.privateKey) {
      throw new Error("account sign only support local model.");
    }

    const h = this.hashMessage(data);
    return Buffer.from((0, _crypto.makeSigner)(0)(h.toString("hex"), this.privateKey), "hex");
  }

  recover(message, signature, preFixed = false) {
    let bytes = message;

    if (!preFixed) {
      bytes = this.hashMessage(message);
    } // @ts-ignore


    return (0, _crypto.recover)(bytes, signature);
  }

  hashMessage(data) {
    let bytes = data;

    if (typeof data === "string" && (0, _utils.isHexStrict)(data)) {
      bytes = (0, _utils.hexToBytes)(data);
    } // @ts-ignore


    const messageBuffer = Buffer.from(bytes);
    const preamble = `\x16IoTeX Signed Message:\n${bytes.length}`;
    const preambleBuffer = Buffer.from(preamble);
    const iotexMessage = Buffer.concat([preambleBuffer, messageBuffer]);
    return (0, _hash.hash256b)(iotexMessage);
  }

}

exports.Account = Account;

class RemoteAccount extends Account {
  constructor(address, sp) {
    super();

    _defineProperty(this, "address", void 0);

    _defineProperty(this, "privateKey", void 0);

    _defineProperty(this, "publicKey", void 0);

    _defineProperty(this, "sp", void 0);

    this.address = address;
    this.sp = sp;
  }

  async sign(data) {
    if (!this.sp.signMessage) {
      return new Buffer("");
    }

    return this.sp.signMessage(data);
  }

}

exports.RemoteAccount = RemoteAccount;
